Allow creating BigNumber with {s, e, c}

iamdoron opened the pull request:
Hi,

`BigNumber` is not yet stable (and might change forever), but I think we can define `BigNumberData` which is pretty stable:

```typescript
  export interface BigNumberData {
    /**
     * The coefficient of the value of this BigNumber, an array of base 1e14 integer numbers.
     */
    readonly c: number[];

     /**
     * The exponent of the value of this BigNumber, an integer number, -1000000000 to 1000000000.
     */
    readonly e: number;

     /**
     * The sign of the value of this BigNumber, -1 or 1.
     */
    readonly s: number;
  }
```

That will actually allow multiple packages to have different BigNumber versions, but still interact *safely* without converting to string. Nice thing about this way is that if you have `BigNumber7` & `BigNumber8`, 
```javascript
new BigNumber8(new BigNumber7('123'))
new BigNumber7(new BigNumber8('123'))
```
just works (presumably safely, even with breaking changes in BigNumber v8 that didn't change BigNumberData).


Maybe we can also version this `BigNumberData` if you believe it will change in the future.
I didn't do it in this PR , but for example:
```typescript
  export interface BigNumberData {
    /**
     * The coefficient of the value of this BigNumber, an array of base 1e14 integer numbers.
     */
    readonly c: number[];

     /**
     * The exponent of the value of this BigNumber, an integer number, -1000000000 to 1000000000.
     */
    readonly e: number;

     /**
     * The sign of the value of this BigNumber, -1 or 1.
     */
    readonly s: number;
     /**
     * The version of this BigNumber representation 
     */
    readonly ver: number
  }
```

and once ver is changed BigNumber will throw when trying to construct with unknown versions

MikeMcl commented:
Hi,
I'll have a look at this properly later this week.

jcramer commented:
Funny, I was coming here to create a similar request.

We are storing our application state as documents to mongoDB and we have a lot of use of BigNumber.  When the documents are saved to Mongo bignumber objects are recorded as `{ s: number, e: number, c: [number], "_isBigNumber": bool }` within the database.

Having a BN constructor that allows creation from this would be hugely helpful to us when restoring our state from the db documents.  

i.e., we need something like: `new BigNumber({ s: number, e: number, c: [number], "_isBigNumber": bool })`

MikeMcl commented on bignumber.d.ts:
```diff
@@ -330,10 +330,28 @@ export namespace BigNumber {
+     */
+    readonly s: number;
+
+    [key: string]: any;
```
Does this stop the object having a symbol as a key? If so, that would be a [problem](https://github.com/MikeMcl/bignumber.js/blob/master/bignumber.mjs#L2695).

iamdoron replied:
according to TS1023
```
An index signature parameter type must be 'string' or 'number'.  ts(1023)
```

It looks like [it's ok](https://www.typescriptlang.org/play/index.html#src=interface%20IInterface%20%7B%0D%0A%20%20num%3A%20number%2C%0D%0A%20%20%5Bkey%3A%20string%5D%3A%20any%0D%0A%7D%0D%0A%0D%0Aconst%20x%3A%20IInterface%20%3D%20%7B%0D%0A%20%20%20%20num%3A%2012%2C%0D%0A%20%20%20%20x%3A%20'string'%2C%0D%0A%20%20%20%20%5BSymbol.toStringTag%5D%3A%20'BigNumber'%0D%0A%7D%0D%0A%0D%0Aconsole.log(x)%0D%0A%0D%0Aconst%20keys%20%3D%20Object.keys(x)%0D%0Aconsole.log(keys%5B0%5D)).

MikeMcl replied:
Looks like it's still an [issue](https://github.com/Microsoft/TypeScript/issues/1863), so probably best without it.

iamdoron replied:
but then it won't work with old BigNumber versions (when `instanceof` doesn't work).

not sure I understand, will you rewrite BigNumber in typescript? 
what I wanted was to allow unknown props, as it will be much easier to construct BigNumber from it. 

for example in js you will be able to do
```javascript
new BigNumber({s: null, e: null, c: null, userSpecific: 'v10' })
```

but in typescript, if we remove the `[key: string]: any`:

```typescript
new BigNumber({s: null, e: null, c: null, userSpecific: 'v10' } as BigNumber.Object)
```

MikeMcl replied:
>> but then it won't work with old BigNumber versions (when `instanceof` doesn't work).

Sorry, I'm not following you. Why would it not work?  (Incidentally, old BigNumber versions have always worked: the constructor just calls `toString` on them.)

MikeMcl commented:
You have defined the ` BigNumberData` interface within the  `BigNumber` namespace, so 

```js
BigNumber.BigNumberData
```
would be added to the existing exports:

```js
//   class      BigNumber  (default export)
//   type       BigNumber.Constructor
//   type       BigNumber.Instance
//   type       BigNumber.ModuloMode
//   type       BigNumber.RoundingMOde
//   type       BigNumber.Value
//   interface  BigNumber.Config
//   interface  BigNumber.Format
```

I would prefer

```js
export interface Object {
```

so the export would be

```js
BigNumber.Object
```

whicih seems more in keeping.

The alternative is to move the interface out of the `BigNumber` namespace, but, on balance, I'd prefer to keep all the exported symbols within it.

I don't want any version property.

@jcramer 

I'm not sure about requiring the `"_isBigNumber"` property, as in *bignumber.mjs* (the ES module version of *bignumber.js*) it doesn't exist (a Symbol is used instead).

MikeMcl commented on bignumber.js:
```diff
@@ -195,7 +195,8 @@
         // Duplicate.
-        if (n instanceof BigNumber) {
+        if (n instanceof BigNumber 
+          || (n != null && n.s != null && n.e != null && n.c != null)) {
```
BigNumber instances with the value `NaN` or `Infinity` have `null` as a property value, so 

```js
|| (n && typeof n.s !== 'undefined' && typeof n.e !== 'undefined' && typeof n.c !== 'undefined')) {
```

iamdoron commented:
@MikeMcl 
- I renamed interface `BigNumber.BigNumberData` to `BigNumber.Object`. It does look better. I though Object was a used keyword, but inside of a namespace it's OK.
- Fixed the `NaN` issue

MikeMcl commented:
There is also BigNumber Infinity which is `{c: null, e: null, s: 1}`.

My preference is:

```js
// strictNullChecks recommended
interface Object {
    readonly c: number[] | null;
    readonly e: number | null;
    readonly s: number | null;
  }
```

On balance, I prefer not to have the `[key:string]: any;`, although I can see its appeal in terms of flexibility, so it is a very close call.


No need to export `BigNumber.Object`.
 `BigNumber.Value` is enough.

```js
export type Value = string | number | BigNumber | Object;
```

MikeMcl commented:
Thanks, Doron.

*v8.1.0* released.

I went with the following in the end:

```js
export interface Instance {
  readonly c: number[] | null;
  readonly e: number | null;
  readonly s: number | null;
}

export type Value = string | number | BigNumber | Instance;
```

To create a BigNumber from an object:

```js
const x = new BigNumber({ s: 1, e: 2, c: [ 777, 12300000000000 ], _isBigNumber: true });
```

To check that a BigNumber instance is well-formed:

```js
BigNumber.DEBUG = true;
BigNumber.isBigNumber(x);    // true
```

See [isBigNumber](http://mikemcl.github.io/bignumber.js/#isBigNumber).

reslake commented:
json2.js
